Remove order::Ordering keyword argument to sort* methods where possible

nalimilan opened the pull request:
This argument is only used internally and was not supposed to be documented
(which was the case in Julia 0.6). Remove it everywhere it is not needed, that
is in all functions except  `sortperm()`, for which that argument is used by
`sortrows()` and `sortcols()` internally. Remove it for docstrings anyway.

Also remove the initialized keyword argument to sort(), which had no effect
and was not documented in Julia 0.6 either.

-------------
Fixes https://github.com/JuliaLang/julia/issues/22382 in the most radical way.

I must say I don't really understand the role of `Ordering` but since @StefanKarpinski said it was supposed to be internal I removed it from all places where it wasn't used. If that's not OK, I can change the PR to remove only the mentions in the docstrings.

Cc @kshyatt who added mentions of these arguments to the docstrings in https://github.com/JuliaLang/julia/pull/18041, probably to match the actual signature.

iamed2 commented:
DataStructures.jl uses Ordering but doesn't use it with `sort`.

StefanKarpinski commented:
Ordering was introduced when we specialized functions on argument types and functions did not have different types. So in order for common sort orders to be fast, they needed to be represented as types with methods defined for the `lt` function, which would be specialized on, type inferred, inlined, etc. We no longer need it for that, so we could get rid of the Ordering types altogether and just pass an `lt` function to the core sorting algorithms. Ordering types could remain useful for something, but I'm not sure there's ever any reason you couldn't just use an `lt` function in its place.

nalimilan commented:
@StefanKarpinski Good to merge then?

tkelman commented:
would be safer to deprecate this than remove it outright

kmsquire commented:
I think it would be good to remove this at the same time that `Orderings` themselves are removed (which could happen right now, if anyone had the time).  Even if not documented, there's always the chance that someone is using Orderings, and it would be odd for it to break here and not everywhere else at the same time.

nalimilan committed:
Deprecate order::Ordering keyword argument to sort* methods

This argument is only used internally and was not supposed to be documented
(which was the case in Julia 0.6). Deprecate it everywhere in the public API,
but keep using it internally for now since the logic is needed to keep
the deprecation working, and for some internal uses like in sortrows() and
sortcols(), which call sortperm() with a Lexicographic ordering.
Remove order for docstrings.

Also deprecate the initialized keyword argument to sort(), which had no effect
and was not documented in Julia 0.6 either.

nalimilan commented:
I've updated the PR to deprecate the arguments rather than removing them. It's not pretty, but...

I've also deprecated for `sortperm` by using an internal `_sortperm` function, to be called by `sortrows` and `sortcols`. That way, `Ordering` is completely absent fro the public API and can be removed when we want.

tkelman commented on base/sort.jl:
```diff
@@ -689,12 +693,14 @@ function sortperm(v::AbstractVector;
 """
-    sortperm!(ix, v; alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward, initialized::Bool=false)
+    sortperm!(ix, v;
+              alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity,
```
will this render right across multiple lines?

nalimilan replied:
Not sure what's the question, but it looks exactly the same at the REPL (and in blue):
```
help?> sortperm!
search: sortperm! sortperm

  sortperm!(ix, v;
            alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity,
            rev::Bool=false, initialized::Bool=false)

```

tkelman replied:
in the html docs - I thought multiline signatures didn't come across right

nalimilan replied:
Ah. Maybe it has improved, but now it looks just like it should -- and quite nicer than long lines, which are hidden and require horizontal scrolling to be read.

nalimilan commented:
Bump.

tkelman commented on base/deprecated.jl:
```diff
@@ -1543,6 +1543,12 @@ end
 @deprecate cat_t{N,T}(::Type{Val{N}}, ::Type{T}, A, B) cat_t(Val(N), T, A, B) false
 @deprecate reshape{N}(A::AbstractArray, ::Type{Val{N}}) reshape(A, Val(N))
 
+# PR #22388: also remove corresonding code in ordering.jl and sort.jl
```
corresponding

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22388#issuecomment-318840268):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/02626235bac61d22ec927c262c55d43e34964a30/b8f0df1_vs_0d7bbcc/report.md)
cc @ararslan

JeffBezanson commented:
There might be uses for having "inspectable" order objects. For example, if you want to sort by a certain key but can only pass a comparison function `(x, y)->isless(x.a, y.a)`, it's hard to determine that we're sorting by `x->x.a`. Column stores would want that information.

nalimilan commented:
Though it looks like you could just define a `a_is_less` function and pass as `isless`? You could even use a callable type `OrderByA` for that. Basically, it seems that `order` is redundant with the other parameters.

nalimilan commented:
Bump. What's the way forward here?

StefanKarpinski commented:
I was convinced that we should remove the `Ordering` type and all the surrounding machinery altogether, but after some discussion with @JeffBezanson, we decided it was often quite useful to have a structured representation of the ordering predicate, even though anonymous functions can be used. Perhaps rather than removing or hiding this we should make it official and full-featured. Currently `by`, `lt` and `rev` all compose in the sense that you can supply them all independently with useful meanings. E.g. `by=f` and `lt=g` means that you sort by `(x,y) -> g(f(x), f(y))`. What if we deprecate `order` as a keyword and only allow an `Ordering` object as an extra argument since you can either supply an explicit `Ordering` argument or implicitly construct one via `lt`, `by` and `rev` keyword arguments – but not both.

StephenVavasis commented:
Here is an argument for keeping the `Ordering` type,  For the sorted containers in DataStructures.jl, the user has the possibility of defining both `lt(o,a,b)` (to check whether `a<b` in ordering `o`) and `eq(o,a,b)` (to check whether `a==b` in ordering `o`).  If the latter is undefined, then equality-testing defaults to `!(a<b) && !(b<a)`.  But an explicit definition of an equality function may be up to twice as fast as the alternative in the case of a lexicographic-style comparison.  The ordering type lets these two functions be grouped together; the alternative of providing two hooks for every sorted-container constructing operation, one for a less-than function and another for an equals-function, would lead to less cohesive code.

nalimilan commented:
I had kind of forgotten this, but we need to make a decision before 0.7. Is everybody fine with the approach @StefanKarpinski proposed?

> What if we deprecate order as a keyword and only allow an Ordering object as an extra argument since you can either supply an explicit Ordering argument or implicitly construct one via lt, by and rev keyword arguments – but not both.

Keno commented:
Triage discussion: It seems a little late to do this now, and we might need a bit of a larger design discussion on what exactly sort should look like and how it fits in with `searchsorted`, etc. It was also pointed out adding the version with the positional argument is technically non-breaking, so we could do it as long as we keep the keyword argument version around (even if deprecated) until the next breaking release.

vtjnash commented:
IIUC, there are good reasons to keep the status quo (e.g. https://github.com/JuliaLang/julia/pull/22388#issuecomment-327277511 above), which means there are reasonable ways that all three keywords may be useful and meaningful in various combinations to various algorithms. They are not entirely independent, but that doesn't mean they aren't usable in combination.
